### PR TITLE
chore: when retrieving a configuration value, check locked and managed-by values

### LINKED ISSUE #13897: Admin config: Compare locked vs defaults vs user config and use accordingly.
### Is your enhancement related to a problem? Please describe

Similar to #13894 and #13896

We want to use the appropriate user config setting.

This would "ideally" be done by a "proxy" for the user config (cc @benoitf for more info)

Which will compare the "admin" config vs user configuration and give back the values accordingly.

### Describe the solution you'd like

A proxy class which will be used for reading configuration which will choose / give values back accordingly to locked / defaults / user config.

Ideally we would **NOT** fail if trying to "save" the values, but instead log / warn somewhere else that we're not able to save due to admin "locked" configuration

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by cdrage**
chore: when retrieving a configuration value, check locked and managed-by values

### What does this PR do?

When retrieving a configuration value from within Podman Desktop API, if
that value is "locked" by the "locked.json' file, then it will:

1. Check if the value exists in the locked.json file
2. Check the "default-settings" json file
3. Return the default-settings json file

If there is no value in the default-settings file (went missing or
something, despite the locked.json existing), then it will return the
users default value, which is intentional, as we do not want to return
null / undefined values just because the value doesn't exist.

There is little code added, but A LOT of tests added in order to cover
every possible edge case.

### Screenshot / video of UI

N/A, internal usage when using locked.json and default-settings.json
files.

### What issues does this PR fix or reference?

Closes https://github.com/podman-desktop/podman-desktop/issues/13897

### How to test this PR?

- [X] Tests are covering the bug fix or the new feature
Signed-off-by: Charlie Drage <charlie@charliedrage.com>

### CODE REVIEW

**Comment by cdrage** [comment]
# SPECIAL NOTE

Please review commit https://github.com/podman-desktop/podman-desktop/pull/14616/commits/eca3f01c89c1246c4e87569b2bca6022b02dc59a from this PR :)

This is built on the work in this PR: https://github.com/podman-desktop/podman-desktop/pull/14316

**Review by jeffmaury — COMMENTED** [review]
Seems duplicate of #14316

**Comment by cdrage** [comment]
> Seems duplicate of #14316

Yup, it's rebased on it please review commit https://github.com/podman-desktop/podman-desktop/pull/14616/commits/7b5376851b81e5d0c833fc2e7150acbdcf3bd0e2

File: packages/main/src/plugin/configuration-impl.ts [hunk 1]
```diff
@@ -54,14 +55,72 @@ export class ConfigurationImpl implements containerDesktopAPI.Configuration {
   get<T>(section: string, defaultValue?: unknown): T | undefined {
     const localKey = this.getLocalKey(section);
 
-    // now look if we have this value
+    // First thing!
```
Comment (axel7083) [inline packages/main/src/plugin/configuration-impl.ts]: this is not required

File: packages/main/src/plugin/configuration-impl.ts [hunk 2]
```diff
@@ -54,14 +55,72 @@ export class ConfigurationImpl implements containerDesktopAPI.Configuration {
...
+    // from managed defaults or locked scopes would be counter-intuitive... (or Onboarding scope, but that doesn't make sense either)
+    if (this.scope === CONFIGURATION_DEFAULT_SCOPE) {
+      // Pass in the key we are wanting to get and check locked config for the key
+      const managedValue = this.getLockedConfigValue<T>(localKey);
+
+      // Before we go ahead and return the value, we want to log to console
+      // this information so that if users are confused why they can't change a setting, they know why,
+      // however, this information should only stay in console.debug considering that it could be noisy (returning configuration values multiple times)
+      //
+      // FUTURE CHANGES:
```
Comment (axel7083) [inline packages/main/src/plugin/configuration-impl.ts]: if this is planed, please link the issue in the comment

File: packages/main/src/plugin/configuration-impl.spec.ts [hunk 3]
```diff
@@ -224,3 +225,444 @@ test('if there are multiple scopes with different values, make sure we get the r
   const resultAdmin = configAdmin.get<string>('key');
   expect(resultAdmin).toBe('adminValue');
 });
+
+// START OF TESTS OF "PICKING" via managed defaults when a value is "locked"
```
Comment (axel7083) [inline packages/main/src/plugin/configuration-impl.spec.ts]: please don't use uppercase in comments, and groups tests using `describe`

File: packages/main/src/plugin/index.spec.ts [hunk 4]
```diff
@@ -68,6 +69,12 @@ class TestPluginSystem extends PluginSystem {
       } as unknown as DefaultConfiguration;
       container.bind<DefaultConfiguration>(DefaultConfiguration).toConstantValue(defaultConfigurationMock);
     }
+    if (!container.isBound(LockedConfiguration)) {
```
Comment (axel7083) [inline packages/main/src/plugin/index.spec.ts:72]: remark: I am not sure to understand the isBound here? why is it needed?

File: packages/main/src/plugin/locked-configuration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+let lockedConfiguration: LockedConfiguration;
+const getManagedDefaultsDirectoryMock = vi.fn();
+const directories = {
+  getManagedDefaultsDirectory: getManagedDefaultsDirectoryMock,
+} as unknown as Directories;
+
+beforeEach(() => {
+  vi.resetAllMocks();
+  vi.clearAllMocks();
```
Comment (axel7083) [inline packages/main/src/plugin/locked-configuration.spec.ts:37]: resetAllMocks is covering the clearAllMocks logic

File: packages/main/src/plugin/locked-configuration.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import * as fs from 'node:fs';
+import * as path from 'node:path';
```
Comment (axel7083) [inline packages/main/src/plugin/locked-configuration.ts]: Please only import what you need, and use `node:fs/promises` for async IO

**Review by axel7083 — COMMENTED** [review]
Left some nit pick 👍

File: packages/main/src/plugin/index.spec.ts (hunk 4 above)
Comment (cdrage) [reply packages/main/src/plugin/index.spec.ts:72]: Similar to

```ts
    if (!container.isBound(DefaultConfiguration)) {
      const defaultConfigurationMock = {
        getContent: vi.fn().mockResolvedValue({}),
      } as unknown as DefaultConfiguration;
      container.bind<DefaultConfiguration>(DefaultConfiguration).toConstantValue(defaultConfigurationMock);
    }
 ```
 
 above as well as configuration registry.
 
 Using it so we can properly mock `getContent` when testing to a constant value / it tries not to "retrieve" it from the filesystem in the mock (just provided the {} content back)

**Comment by cdrage** [comment]
Thanks! I've updated the code.

For the changes to locked configuration, I'll apply your suggestions here: https://github.com/podman-desktop/podman-desktop/pull/14316

**Review by bmahabirbu — APPROVED** [review]
LGTM! Ran through the test locally on mac all good

File: packages/main/src/plugin/configuration-impl.ts [hunk 7]
```diff
@@ -54,14 +55,70 @@ export class ConfigurationImpl implements containerDesktopAPI.Configuration {
...
+      // Before we go ahead and return the value, we want to log to console
+      // this information so that if users are confused why they can't change a setting, they know why,
+      // however, this information should only stay in console.debug considering that it could be noisy (returning configuration values multiple times)
+      //
+      // A GUI indication will be provided in the settings page for locked values:
+      // https://github.com/podman-desktop/podman-desktop/issues/13900
+      if (managedValue !== undefined) {
+        console.debug(
+          `[Managed-by]: Configuration key '${localKey}' is locked. Returning managed default value instead of user value.`,
+        );
```
Comment (benoitf) [inline packages/main/src/plugin/configuration-impl.ts]: IMHO there are many comments that will be unmaintained and that should not belong there
we don't need to know in configuration-impl for example
```
// We retrieve the information from from a system file:
// (ex macOS: /Library/Application Support/com.podman.desktop/locked.json, Linux: /usr/share/podman-desktop/locked.json)
```

all of that is part of how LockedConfig works, not ConfigurationRegistry

File: packages/main/src/plugin/configuration-impl.ts [hunk 8]
```diff
@@ -54,14 +55,70 @@ export class ConfigurationImpl implements containerDesktopAPI.Configuration {
...
+
+    // This key IS locked - grab the managed default value for it
+    const managedDefaults = this.configurationValues.get(CONFIGURATION_SYSTEM_MANAGED_DEFAULTS_SCOPE);
+    if (managedDefaults?.[localKey] !== undefined) {
+      return managedDefaults[localKey] as T;
+    }
+
+    return undefined;
+  }
+
```
Comment (benoitf) [inline packages/main/src/plugin/configuration-impl.ts]: it should not be there, it should be in a LockConfiguration object

const managedValue = this.getLockedConfigValue<T>(localKey);
-->

this.lockedConfiguration.get(...) and it's in this class that you can write comments about lock configuration

File: packages/main/src/plugin/locked-configuration.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+import { inject, injectable } from 'inversify';
+
+import { Directories } from './directories.js';
+import { SYSTEM_LOCKED_FILENAME } from './managed-by-constants.js';
+
+@injectable()
+export class LockedConfiguration {
```
Comment (benoitf) [inline packages/main/src/plugin/locked-configuration.ts:28]: IHMO it's confusing that we have a LockedConfiguration that read the locked file and all documentation is in configuration-impl while it applied here

and why getLockedConfigValue do not belong to this object ?

File: packages/main/src/plugin/configuration-impl.ts [hunk 10]
```diff
@@ -54,14 +55,70 @@ export class ConfigurationImpl implements containerDesktopAPI.Configuration {
...
+      const managedValue = this.getLockedConfigValue<T>(localKey);
+
+      // Before we go ahead and return the value, we want to log to console
+      // this information so that if users are confused why they can't change a setting, they know why,
+      // however, this information should only stay in console.debug considering that it could be noisy (returning configuration values multiple times)
+      //
+      // A GUI indication will be provided in the settings page for locked values:
+      // https://github.com/podman-desktop/podman-desktop/issues/13900
+      if (managedValue !== undefined) {
+        console.debug(
```
Comment (benoitf) [inline packages/main/src/plugin/configuration-impl.ts]: it will produce way too many log, IMHO should be removed

Also the [Managed-By] is very confusing as it's dispatched in too many locations.

**Review by benoitf — CHANGES_REQUESTED** [review]

**Comment by cdrage** [comment]
Updated the PR cc @benoitf 

Thanks for the review! It's now in a separate class, the comments are less verbose, and I made it way less confusing (did your suggestion of moving the method to the other class).

Ready for another look (https://github.com/podman-desktop/podman-desktop/pull/14616/commits/828a37f98616ef66993f25bc81c79b2e3929c487)

File: packages/main/src/plugin/locked-configuration.ts (hunk 9 above)
Comment (cdrage) [reply packages/main/src/plugin/locked-configuration.ts:28]: I agree, changed over the comments so it's less verbose (explained too much) and moved the comments over so it's easier to understand

File: packages/main/src/plugin/configuration-impl.ts (hunk 10 above)
Comment (cdrage) [reply packages/main/src/plugin/configuration-impl.ts]: ACK, I agree

File: packages/main/src/plugin/configuration-impl.ts (hunk 8 above)
Comment (cdrage) [reply packages/main/src/plugin/configuration-impl.ts]: moved it over to a different class now

File: packages/main/src/plugin/configuration-registry.spec.ts [hunk 11]
```diff
@@ -49,6 +49,16 @@ vi.mock('node:fs', () => ({
   },
 }));
 
+// mock DefaultConfiguration for the new managed defaults functionality
+vi.mock('./default-configuration.js', () => ({
```
Comment (axel7083) [inline packages/main/src/plugin/configuration-registry.spec.ts]: nitpick: use the import synthax, this ensure that the import exists, and allow typecheck to validate it


```suggestion
vi.mock(import('./default-configuration.js'), () => ({
```

File: packages/main/src/plugin/configuration-registry.spec.ts [hunk 12]
```diff
@@ -49,6 +49,16 @@ vi.mock('node:fs', () => ({
...
+  DefaultConfiguration: vi.fn(),
+}));
+
+// mock LockedConfiguration for the new managed locked functionality
+vi.mock('./locked-configuration.js', () => ({
```
Comment (axel7083) [inline packages/main/src/plugin/configuration-registry.spec.ts]:
```suggestion
vi.mock(import('./locked-configuration.js'), () => ({
```

**Review by axel7083 — APPROVED** [review]
LGTM, left two suggestions

File: packages/main/src/plugin/lock-configuration.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+   * of the user-defined one.
+   *
+   * @param localKey - Configuration key to check
+   * @returns return the managed value if locked, undefined otherwise
+   */
+  get<T>(localKey: string): T | undefined {
+    const lockedConfig = this.configurationValues.get(CONFIGURATION_SYSTEM_MANAGED_LOCKED_SCOPE);
+
+    // Bail early if there's no locked config or it's malformed
+    if (!lockedConfig?.['locked'] || !Array.isArray(lockedConfig['locked'])) {
```
Comment (simonrey1) [inline packages/main/src/plugin/lock-configuration.ts]: suggestion(non-blocking): you can also safeParse the value with zod library. If you have a result, you can continue, else you return undefined (like on l. 42).

File: packages/main/src/plugin/configuration-registry.spec.ts (hunk 12 above)
Comment (cdrage) [reply packages/main/src/plugin/configuration-registry.spec.ts]: is there a lint setting for this? cause I do this way too much 🤣

**Review by jeffmaury — APPROVED** [review]
LGTM

File: packages/main/src/plugin/lock-configuration.ts (hunk 13 above)
Comment (cdrage) [reply packages/main/src/plugin/lock-configuration.ts]: Opened up https://github.com/podman-desktop/podman-desktop/issues/14774 for discussion, using zod looks interesting!

**Review by benoitf — APPROVED** [review]
